### ISSUE TITLE
Tracking issue • Role-based authorization system

### ISSUE THREAD

**Issue by leovalais** [OpenRailAssociation/osrd#8171]
> [!NOTE]
> Design doc: https://osrd.fr/en/docs/reference/design-docs/auth/

# Setup and logic system — covered in #7981 

- [X] Crate `editoast_authz`
- [X] SQL setup
- [X] New user registration using request headers
- [X] Role config
- [X] Internal API setup (`Authorizer` and friends)
- [X] Role check
- [X] Test over one endpoint
- [X] Works with manual role edition (no endpoint, manual DB modifications)
- [x] Logic tests on `editoast_authz` side using mocked DB interface
- [X] DB interface tests
- [ ] Some documentation ofc

# Planned work

- [ ] https://github.com/OpenRailAssociation/osrd/issues/8743
- [ ] https://github.com/OpenRailAssociation/osrd/issues/8744
- [ ] https://github.com/OpenRailAssociation/osrd/issues/8746

# Various improvements

- [ ] Make a middleware that checks the roles of each endpoints automatically (we could hook ourselves to the `struct __path_fn` generated for each endpoint to provide the required roles) ✋ Blocked by https://github.com/OpenRailAssociation/osrd/issues/8178 (item `routes!`)
- [ ] Nice to have but useful: have all protected endpoints take a `Connection` extractor which gets a connection from the pool and starts a transaction. This way we ensure that nothing done before checking the roles/grants can't be undone. ✋ Blocked by https://github.com/OpenRailAssociation/osrd/issues/6980
- [x] #8276
- [x] #8277
